// @expo/xdl/build/tools/ModuleVersion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createModuleVersionChecker = createModuleVersionChecker;

var _latestVersion = _interopRequireDefault(require("latest-version"));

var _pTimeout = _interopRequireDefault(require("p-timeout"));

var _semver = _interopRequireDefault(require("semver"));

var _FsCache = require("./FsCache");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createModuleVersionChecker(name, currentVersion) {
  const UpdateCacher = new _FsCache.Cacher(async () => {
    return {
      latestVersion: await (0, _pTimeout.default)((0, _latestVersion.default)(name), 2000)
    };
  }, `${name}-updates.json`, 24 * 60 * 60 * 1000 // one day
  );

  async function checkAsync() {
    const {
      latestVersion
    } = await UpdateCacher.getAsync();
    return {
      updateIsAvailable: _semver.default.gt(latestVersion, currentVersion),
      latest: latestVersion,
      current: currentVersion
    };
  }

  return {
    checkAsync
  };
}
