// @expo/xdl/build/Config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var Env = _interopRequireWildcard(require("./Env"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

let scheme = process.env.XDL_SCHEME || 'https';
let host = process.env.XDL_HOST || 'exp.host';
let port = parseInt(process.env.XDL_PORT, 10) || null;

if (Env.isStaging()) {
  host = 'staging.exp.host';
} else if (Env.isLocal()) {
  scheme = 'http';
  host = 'localhost';
  port = 3000;
}

const config = {
  api: {
    scheme,
    host,
    port
  },
  ngrok: {
    authToken: '5W1bR67GNbWcXqmxZzBG1_56GezNeaX6sSRvn8npeQ8',
    authTokenPublicId: '5W1bR67GNbWcXqmxZzBG1',
    domain: 'exp.direct'
  },
  developerTool: null,
  validation: {
    reactNativeVersionWarnings: true
  },
  helpUrl: 'https://docs.expo.io/',
  offline: false
};
var _default = config;
exports.default = _default;
